lib/SILOptimizer/SemanticARC/CopyValueOpts.cpp

```diff
@@ -452,6 +453,65 @@ bool SemanticARCOptVisitor::tryJoiningCopyValueLiveRangeWithOperand(
   return false;
 }
 
+/// Given an owned value that is completely enclosed within its parent owned
+/// value and is not consumed, eliminate the copy.
+bool SemanticARCOptVisitor::tryPerformOwnedCopyValueOptimization(
+    CopyValueInst *cvi) {
+  if (ctx.onlyGuaranteedOpts)
+    return false;
+
+  auto originalValue = cvi->getOperand();
+  if (originalValue.getOwnershipKind() != OwnershipKind::Owned)
+    return false;
+
+  // TODO: Add support for forwarding insts here.
+  SmallVector<DestroyValueInst *, 8> destroyingUses;
+  SmallVector<Operand *, 32> allCopyUses;
+  for (auto *use : cvi->getUses()) {
+    // First just stash our use so we have /all uses/.
+    allCopyUses.push_back(use);
+
+    // Then if we are not a lifetime ending use, just continue.
+    if (!use->isLifetimeEnding()) {
+      continue;
+    }
+
+    // Otherwise, if we have a destroy value lifetime ending use, stash that.
+    if (auto *dvi = dyn_cast<DestroyValueInst>(use->getUser())) {
+      destroyingUses.push_back(dvi);
+      continue;
+    }
+
+    // Otherwise, just bail for now.
+    return false;
+  }
+
+  // NOTE: We do not actually care if the parent's lifetime ends with
+  // destroy_values. All we care is that it is lifetime ending and the use isn't
+  // a forwarding instruction.
+  SmallVector<Operand *, 8> parentLifetimeEndingUses;
+  for (auto *origValueUse : originalValue->getUses())
+    if (origValueUse->isLifetimeEnding() &&
+        !isa<OwnershipForwardingInst>(origValueUse->getUser()))
```

Comment:
Hmm... reading this... I think what I wrote here could be made better. I think at this point, the optimization is not dealing with LiveRanges, so the right thing to do is just gather up all consuming uses of the value, ignoring forwarding.

Reply:
*ignoring forwarding I mean just treating forwarding insts as normal consuming instructions.